### ISSUE TITLE
Equalizer has too little bands.

### ISSUE THREAD

**Issue by dikonov** [DeaDBeeF-Player/deadbeef#1921]
My headphones have irregular frequency responce curve (I have measured it). This issue plagues all headphones except some super-expensive Hi-End ones. The sound quality can be improved greatly by using tailored eq presets.

Existing deabeef EQ is too simple to achieve good results. The number of frequency bands is inadequate for this task. For example it cannot fix a sharp hole at 12000Hz, beacuse there is no 12000 band control and both 10000 & 14000hz frequencies are OK and should not be changed.  

Could you add more frequency bands (30+ band eq) or, better, implement parametric EQ presets from Audacity.

Deadbeef version: 7.02
OS: Linux

**Comment by Alexey-Yakovenko** [comment]
It's not as simple as taking the code from Audacity.

There are many reasons why this is not possible -- licensing, a different UI toolkit which could not be used in deadbeef, a different programming language, different performance requirements, and many other things.

The feature request itself is fine, but with 240 other feature requests in front of you, I wouldn't count on this to become available anytime soon. It's probably easier to buy another pair of headphones etc.

**Comment by dikonov** [comment]
README at https://github.com/JoeyZheng/deadbeef/tree/master/plugins/supereq  says taht SuperEq can be PARAMETRIC. Is it the same SuperEq already bundled? If yes, there must be a way to load parametric presets created outside of deadbeef. Audacity can be used as a curve editor than.

**Comment by Alexey-Yakovenko** [comment]
I've no idea why you're looking in some unofficial fork for deadbeef source code.. the same folder is right here: https://github.com/DeaDBeeF-Player/deadbeef/tree/master/plugins/supereq

It is indeed parametric. However, I think what you understand under this term is different from what DSP developers mean, including the developer of SuperEQ.

**Comment by dikonov** [comment]
I am looking for another/alternative EQ for Deadbeef but not the player itself. 

The term "parametric equalizer" (as in Audacity  "graphic" vs "parametric" mode) means an EQ, where the curve can be modified at _any_ frequency rather than only at fixed points and independently for different audio channels. Comparing .ddbeq preset files with the audacity XML preset tells the difference as I see it.

**Comment by Alexey-Yakovenko** [comment]
Supereq is not alternative -- that's the one that ships with deadbeef.
I'm not aware of any 3rd party equalizers developed specifically for deadbeef.
You might have some luck with equalizers developed for musicians / DAWs, which plug into JACK, or LASPA, or whatever else is the current trend on Linux.

**Comment by dikonov** [comment]
I have no idea how to use a LADSPA plugin with Deadbeef. It has no LADSPA host in the DSP plugin chain? If supereq can be modified to import "parametric" curves like:
    arbitrary freq point1 - offset
    arbitrary freq point2 - offset
    ...
Even if the number of points is still limited by 18 the problem will be solved.

**Comment by Alexey-Yakovenko** [comment]
ALSA itself is supposed to have LADSPA support.

And yes, it's quite likely that supereq code can be modified to support center frequency adjustment (not sure about bandwidth adjustment though).

It just needs another adjustment table alongside the `bands` in Equ.cpp.

99% of the work is to add this stuff to UI, change the eq file format, implement backwards compat, test it, and make it shippable. Which might not seem like a lot, but unfortunately it is.

And unless someone volunteers to do that work - I'm afraid that you don't have many options.

**Comment by dikonov** [comment]
To hell the EQ UI. 
An alternative patched supereq as a DSP plugin without UI, that would just load parametric presets might solve the problem. 

LADSPA in ALSA is too complicated to use. I tried and failed to make it work before. Besides, there must be a GUI to turn equalization on and off quickly, when the headphones are unplugged. DSP presets in Deadbeef do it.

**Comment by Alexey-Yakovenko** [comment]
Yep.. just hold on while I'm going through the rest of 240 feature requests :-)

**Comment by wdlkmpx** [comment]
Yeah Alexey has done a supreme work maintaining the current supported
plugins and code, and it's way too much for one man alone.

For those who know how to contribute code or have the potential to
contribute something, forget about all the other players and focus on
deadbeef alone, it's the only universal audio player worthy of your
attention and effort..

**Comment by revast** [comment]
You can script that functionality with the use of JACK audio and an LADSPA or LV2 plugin.
That bash script will do the following:

- start JACK (e.g. via qjackctl)
- start DeaDBeeF (JACK output plugin must be selected first, as DeaDBeeF does not automatically recognize audio servers)
- disconnect deadbeef from system:playback (via jack_connect, as it automatically connects on startup)
-          jack_disconnect 'deadbeef:deadbeef_1'  'system:playback_1'
-          jack_disconnect 'deadbeef:deadbeef_2'  'system:playback_2' 
- start your favorite plugin host (jalv for LV2  and jackrack for LADSPA plugins for example), with your favourite eq (like x42-plugins' Eq or Eq10q )
- connect your DeaDBeeF outputs to the inputs plugin, e.g.
-         jack_connect 'deadbeef:deadbeef_1'  'EQ10Q Stereo:input1
-         jack_connect  'deadbeef:deadbeef_2'  'EQ10Q Stereo:input2'
- connect the Plugin's output to your system output:
-        jack_connect 'system:playback_1' 'EQ10Q Stereo:output1'
-        jack_connect 'system:playback_2'  'EQ10Q Stereo:output2'

For knowing what specifically to type for the plugin (like EQ10Q Stereo:...) you load the plugin into the plugin host, and read the corresponding status change messages in the qjackctl messages window.

**Comment by tgpro** [comment]
@dikonov, you may use [bmc0/dsp](/bmc0/dsp) as a system-wide equalizer with pulse or alsa.
[System-Wide-DSP-Guide](https://github.com/bmc0/dsp/wiki/System-Wide-DSP-Guide/) is a way to go.
Не благодари :)

**Comment by nilejsph** [comment]
I've configured the dsp equalizer and it is pretty good.
 https://github.com/bmc0/dsp/wiki/System-Wide-DSP-Guide/

**Comment by revast** [comment]
wow, good find, I am gonna try this!

**Comment by nBalykin** [comment]
A parametric equalizer is really a necessary function for a player with an emphasis on sound.

**Comment by saaj** [comment]
For PulseAudio there's great [PulseEffects][1] which is the EQ suggested by [AutoEq][2] (the project that helps neutralising frequency response of many headphones).

[1]: https://github.com/wwmm/pulseeffects
[2]: https://github.com/jaakkopasanen/AutoEq

**Comment by dikonov** [comment]
> For PulseAudio there's great [PulseEffects](https://github.com/wwmm/pulseeffects) 
Yeah, but Pulseaudio itself is an evil malware that unexpectedly degrades hifi sound quality AND it is getting phased out in favor of :shrug: Pipewire.

**Comment by revast** [comment]
> > For PulseAudio there's great [PulseEffects](https://github.com/wwmm/pulseeffects)
> 
> Yeah, but Pulseaudio itself is an evil malware that unexpectedly degrades hifi sound quality AND it is getting phased out in favor of 🤷 Pipewire.
> 
> _Normally, I try to always have straight bit-perfect PCM transfer from file to external DAC. An EQ is great only with some headphones and should be turned off as soon as they are unplugged, rather than silently work every time you forget to edit obscure settings in a 3rd party piece of software._
> 
> I would suggest to just hide the EQ UI if a "parametric" eq preset is loaded leaving only the EQ "on"/"bypass" button and preset selector. It would be cool to display the curve, though.

project now renamed to [easyeffects ](https://github.com/wwmm/easyeffects) and targeting pipewire.

**Comment by vkx86** [comment]
Since most headphone listeners who're aware of EQ uses [AutoEq](https://autoeq.app/),
an elegant solution is to create 18-band Custom Parametric EQ and use the values for Deadbeef's SuperEQ.
That's 5-munute task using their [instructions](https://github.com/jaakkopasanen/AutoEq/wiki/Choosing-an-Equalizer-App#non-standard-graphic-equalizers).

@Oleksiy-Yakovenko: From a quick look into code I did, it seems SuperEQ can be configured for 31 bands that have same frequencies as in AutoEQ. That can be perfect solution IMO.

**Comment by Oleksiy-Yakovenko** [comment]
@vkx86 I doubt it. Probably not.

**Comment by dikonov** [comment]
Thanks, @vkx86!  
The AutoEq site is a nice tool but it isn't as world-famous as you say. Nevertheless it is the only solution that actually wors to some extent. The problem is that custom 18-band parametric EQ settings it can generate still leave great dents and peaks in the mid and high frequencies, e.g. around 2300 Hz. It can lessen the sound distortion introduced by earphones, but cannot fix it. 18-band is too little. A 180-band Eq would be a better approximation.
  
Easyeffects is completely useless because it relies on pulse/pipewire, which in turn is an unwelcome middleman. Besides pipewire does not work in my system. I get only silence through it, while both alsa and pulse work well.

**Comment by vkx86** [comment]
@dikonov , 
In **Head-Fi**, AutoEq is a starting point for folks, who want to fine-tune their cans. 
Totally agree - nothing can beat Parametric EQ, especially for headphones.
I'd compared MathAudio Headphone EQ for Foobar2K with same 18-band Graphic EQ used both in F2K and Deadbeef - Parametric EQ wins.
At end, when I do work on Ubuntu - I'm using Audacious, that has LADSPA host **within** the app with PEQ LADSPA plugin and with some Python script to load configurations to configure it for my 2 cans.
